registry/registry_validation.go

```diff
@@ -0,0 +1,167 @@
+			// permitting the action (assuming their keys haven't been stolen!)
+			matched, err := matchKeys(pubkey, superspaces)
+			if err != nil {
+				serverError = errors.Errorf("%v: Unable to check if the incoming key for %s matched any public keys for %s", err, prefix, subspaces)
+				return
+			}
+			if !matched {
+				validationError = errors.New("Cannot register a namespace that is suffixed or prefixed by an already-registered namespace unless the incoming public key matches a registered key")
+				return
+			}
+
+		} else if len(subspaces) > 0 {
+			// If there are no superspaces, we can check the subspaces.
+
+			// TODO: Eventually we might want to check only the highest level subspaces and use those keys for matching. For example,
+			// if /foo/bar and /foo/bar/baz are registered with two keysets such that the complement of their intersections is not null,
+			// it may be the case that the only key we match against belongs to /foo/bar/baz. If we go ahead with registration at that
+			// point, we're essentially saying /foo/bar/baz, the logical subspace of /foo/bar, has authorized a superspace for both.
+			// More interestingly, if /foo/bar and /foo/baz are both registered, should they both be consulted before adding /foo?
+
+			// For now, we'll just check for any key match.
+			matched, err := matchKeys(pubkey, subspaces)
+			if err != nil {
+				serverError = errors.Errorf("%v: Unable to check if the incoming key for %s matched any public keys for %s", err, prefix, subspaces)
+				return
+			}
+			if !matched {
+				validationError = errors.New("Cannot register a namespace that is suffixed or prefixed by an already-registered namespace unless the incoming public key matches a registered key")
+				return
+			}
+		}
+	}
+	return
+}
+
+func validateJwks(jwksStr string) (jwk.Key, error) {
+	clientJwks, err := jwk.ParseString(jwksStr)
+	if err != nil {
+		return nil, errors.Wrap(err, "Couldn't parse the pubkey from the request")
+	}
+
+	if log.IsLevelEnabled(log.DebugLevel) {
+		// Let's check that we can convert to JSON and get the right thing...
+		jsonbuf, err := json.Marshal(clientJwks)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to marshal the reuqest pubKey's keyset into JSON")
+		}
+		log.Debugln("Client JWKS as seen by the registry server:", string(jsonbuf))
+	}
+
+	/*
+	 * TODO: This section makes the assumption that the incoming jwks only contains a single
+	 *       key, a property that is enforced by the client at the origin. Eventually we need
+	 *       to support the addition of other keys in the jwks stored for the origin. There is
+	 *       a similar TODO listed in client_commands.go, as the choices made there mirror the
+	 *       choices made here.
+	 */
+	key, exists := clientJwks.Key(0)
+	if !exists {
+		return nil, errors.New("There was no key at index 0 in the reuqest pubKey's JWKS. Something is wrong")
+	}
+	return key, nil
+}
+
+func validateInstitution(inst string) bool {
```

Comment:
Fairly large change requested:

Please map the institution name to an internal identifier and return the internal identifier here.

For example, the following are entries in the OSG institutional spreadsheet:

```
OSG Identifier                    | Institution Name
--------------------------------- | ---------------------
https://osg-htc.org/iid/05bxb3784 | Academia Sinica
https://osg-htc.org/iid/02d2m2044 | Agricultural Research Service
https://osg-htc.org/iid/05cf8a891 | Albert Einstein College of Medicine
https://osg-htc.org/iid/03thb3e06 | American Museum of Natural History
https://osg-htc.org/iid/00ff4bt20 | Arcadia University
```

We should always display the "Institution name" to the user but, internally, serialize the identifier into the database.  That implies that the `Registry.Institutions` parameter is more complicated than a string slice.

Longer-term (separate PR), we should provide the option of either listing all of the ID<->Institution mappings in the configuration file or providing an API that can be queried.  When running in "OSDF mode", the API would be automatically set to a relevant topology URL.